src/RawPOSIX/src/safeposix/dispatcher.rs

```diff
@@ -205,22 +205,24 @@ pub fn lind_syscall_api(
     let ret = match call_number {
         WRITE_SYSCALL => {
             let fd = arg1 as i32;
-            let buf = (start_address + arg2) as *const u8;
             let count = arg3 as usize;
-            interface::cagetable_getref(cageid)
-                .write_syscall(fd, buf, count)
+            let cage = interface::cagetable_getref(cageid);
+            let buf = match check_and_convert_addr_ext(&cage, arg2, count, PROT_WRITE) {
```

Comment:
I believe this is supposed to be PROT_WRITE, because the write syscall is reading from the buffer, before it writes that data to an FD. 

I see a few errors like this in some of these calls, can you double check them all.